pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
+//
+// It applies some function on the received batches and returns an error if one appears on the error channel.
+//
+// Example usage:
+//
+//		err := DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
+//			bundles = append(bundles, bundleBatch...)
+//		})
+func DoSelectBundles(bundlesChan <-chan model.BundleDescriptors, errorChan <-chan error, do func(model.BundleDescriptors)) error {
+	var err error
+
+LOOP:
+	// consume batches of ordered bundle metadata
+	for {
+		select {
+		case bundleBatch, ok := <-bundlesChan:
+			if !ok {
+				// last attempt to catch an error
+				// (i.e. bundlesChan has closed, but an error was also present on errorChan and hasn't been selected)
+				if err == nil {
+					if e, ok := <-errorChan; ok {
+						err = e
+					}
+				}
+				break LOOP
+			}
+			do(bundleBatch)
+		case e, ok := <-errorChan: // do not exit here: the only natural exit occurs when bundlesChan is closed
+			if ok && err == nil { // retain first error
+				err = e
+			}
+		}
+	}
+
+	return err
+}
+
+// ApplyBundleFunc is a function to be applied on a bundle
+type ApplyBundleFunc func(model.BundleDescriptor) error
+
+// ListBundlesApply applies some function to the retrieved bundles, in lexicographic order of keys.
 //
-// NOTE: resulting bundles are returned in no particular order.
+// The execution of the applied function does not block background retrieval of more keys and bundle descriptors.
 //
-// TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]model.BundleDescriptor, error) {
-	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+// Example usage: printing bundle descriptors as they come
+//
+//   err := core.ListBundlesApply(repo, store, func(bundle model.BundleDescriptorOption) error {
+//				fmt.Fprintf(os.Stderr, "%v\n", bundle)
+//				return nil
+//			})
+func ListBundlesApply(repo string, store storage.Store, apply ApplyBundleFunc, opts ...BundleListOption) error {
+	var (
+		err, applyErr error
+		wg            sync.WaitGroup
+		workDone      bool
+		mutex         sync.Mutex
+	)
+
+	bundleChan := make(chan model.BundleDescriptor)
+	doneChan := make(chan struct{}, 1)
+
+	// collect bundle metadata asynchronously
+	wg.Add(1)
+	go func(bundleChan chan<- model.BundleDescriptor, doneChan chan struct{}, wg *sync.WaitGroup) {
+		defer func() {
+			close(bundleChan)
+			wg.Done()
+		}()
+
+		bundlesChan, errChan, workers := ListBundlesChan(repo, store, append(opts, WithBundleDoneChan(doneChan))...)
+
+		err = DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
+			for _, bundle := range bundleBatch {
+				bundleChan <- bundle // transfer a batch of metadata to the applied func
+			}
+		})
+
+		mutex.Lock()
+		workDone = true // nothing left to interrupt
+		close(doneChan)
+		mutex.Unlock()
+
+		workers.Wait()
```

Comment:
again, prefer waiting on channels (with this `for range c {}` idiom) to a waitgroup when the two are equivalent.

reason:  less mutable state is simpler than less procedural data.

Reply:
ok with the principle but this is not equivalent